lib/hanami/application.rb

```diff
@@ -322,28 +322,26 @@ def initialize(application = self.class)
 
         application.boot
 
-        resolver = application.config.router_endpoint_resolver.new(
+        resolver = application.config.router_resolver.new(
           container: application,
           namespace: application.config.router_endpoint_container_key_namespace,
+          inflector: application.inflector
         )
 
+        stack = Hanami::Application::Routing::Middleware::Stack.new
+        stack.use application[:rack_monitor]
+        application.config.for_each_middleware do |m, *args, &block|
+          stack.use(m, *args, &block)
+        end
+
         router = Application::Router.new(
-          context: application,
-          endpoint_resolver: resolver,
+          resolver: resolver,
+          stack: stack,
           **application.configuration.router_settings,
           &application.routes
         )
 
-        @app = Rack::Builder.new do
-          use application[:rack_monitor]
-
-          # Apply middleware from configuration
-          application.config.for_each_middleware do |m, *args, &block|
-            use(m, *args, &block)
-          end
-
-          run router
-        end
+        @app = stack.finalize(router)
```

Comment:
Now that you've explained the reason for the stack ivar being removed when the router is frozen, this whole arrangement does feel a little bit awkward... i.e. the fact that we:

1. Create a `stack`
2. Create a `router`, passing the `stack` into it
3. And then finalize the `stack`, passing the `router` back to it!

So these two things feel very tightly coupled...

I understand the reason for the stack being passed into the router, so middleware can be used at various route namespaces, but I wonder if there's a way we could clean this relationship up some more? What do you think?

FWIW, I'd be happy just to flag this as something for us to follow up on later, but I felt it was worth mentioning here while it's fresh code.

Reply:
This is for perf reasons.

If `Stack` is empty, the returning value assigned to `@app` is the router itself. Otherwise, the returning value is a `Rack::Builder` instance.

Making all the requests to go to Rack middleware stack has a perf cost that we want to **avoid** if no middleware is set up by a dev.

Reply:
One way to decouple the finalization part would be:

```ruby
@app = if stack.empty?
  router
else
  stack.build
end
```

I've seen an opportunity for a "tell don't ask" approach with `Stack#finalize`. But we can move it to `Middleware.build(router, stack)` and that `if` will be handled by it:

```ruby
@app = Middleware.build(router, stack)
```

What do you think?

Reply:
Yeah, that could help tease things apart a little.

I've another idea to throw into the mix, too: is there any reason our "application router" couldn't just create it's _own_ stack internally, when it's initialized? Then the `router` object is the only thing we deal with from within `Hanami::Application#initialize`, e.g.

```ruby
# This router would have @stack = Stack.new inside it's #initialize
router = Application::Router.new(
  resolver: resolver,
  **application.configuration.router_settings,
)

# These would go straight onto the router's own internal stack
router.use application[:rack_monitor]
application.config.for_each_middleware do |m, *args, &block|
  stack.use(m, *args, &block)
end

# Instead of calling instance_eval directly, we could create a method on
# Application::Router to make this clearer
router.instance_eval(&application.routes)

# Equivalent to stack.finalize
@app = router.to_rack_app
```

How about that?

Reply:
@timriley please have a look at 4204cf6

Reply:
@jodosha I LOVE IT!! 🥰 

That exception you comment about is kind of strange. Do you want me to take a look?

Reply:
@timriley Yes, please. I've tried to reproduce the problem in `dry-monitor` master, but this test passes. Is it possible that `v0.3.1` the latest stable release diverges in behavior from master?

```diff
diff --git a/spec/integration/rack_middleware_spec.rb b/spec/integration/rack_middleware_spec.rb
index d32a0bd..4ca632a 100644
--- a/spec/integration/rack_middleware_spec.rb
+++ b/spec/integration/rack_middleware_spec.rb
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'rack/builder'
+
 RSpec.describe Dry::Monitor::Rack::Middleware do
   subject(:middleware) { Dry::Monitor::Rack::Middleware.new(notifications).new(rack_app) }
 
@@ -40,6 +42,25 @@ RSpec.describe Dry::Monitor::Rack::Middleware do
     ].join('&')
   end
 
+  describe '#new' do
+    let(:builder) do
+      middleware = described_class.new(notifications)
+      app = rack_app
+
+      Rack::Builder.new do
+        use middleware
+        run app
+      end
+    end
+
+    it 'is compatible with Rack::Builder' do
+      expect(rack_app).to receive(:call).with(env).and_return([200, :total_success])
+
+      app = builder.to_app
+      app.call(env)
+    end
+  end
+
   describe '#call' do
     before do
       File.open(log_file_path, 'w').close
```

Reply:
https://github.com/dry-rb/dry-monitor/pull/37